met-api/src/met_api/utils/filter_types.py

```diff
@@ -0,0 +1,43 @@
+"""
+Filters used to filter by metadata contents in various ways.
+"""
+
+from met_api.models.engagement_metadata import EngagementMetadata, MetadataTaxonFilterType
+from sqlalchemy import select, func, distinct, exists
+
+
+from sqlalchemy.sql import select, func, distinct, exists
+
+
+def list_match_all(query, values):
+    # Create a subquery to find engagements that have the required number of matching metadata entries
+    values_count = len(values)
+    subquery = query.filter(
+        EngagementMetadata.value.in_(values)
+    ).group_by(
+        # Group by engagement_id to count the number of matching values
+        EngagementMetadata.engagement_id
+    ).having(
+        # Ensure that the number of matching values is equal to the number of values provided
+        func.count('*') == values_count
+    ) .subquery()
+    return subquery
+
+
+def list_match_any(query, values):
```

Comment:
You may be able to merge these two functions to save a bit of code. Something like:

```suggestion
def list_match(query, values, filter_match_type):
    query = query.filter(
        EngagementMetadata.value.in_(values)
    )
    
    if (filter_match_type == 'all'):
        # append other query operations.
    
    return query.subquery()
```

Reply:
Choosing to keep these separate for now in the interest of time and extensibility.

Note to future self: Alex suggested implementing these as a factory function that is able to return a callback for the engagement filtering engine to use based on a name, instead of using the map.